src/map.rs

```diff
@@ -269,24 +273,37 @@ impl<T> Server<T> {
     }
 
     /// Starts the gRPC server. When message is received on the `shutdown` channel, graceful shutdown of the gRPC server will be initiated.
-    pub async fn start_with_shutdown<F>(
+    pub async fn start_with_shutdown(
         &mut self,
-        shutdown: F,
+        shutdown_rx: oneshot::Receiver<()>,
     ) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
     where
         T: Mapper + Send + Sync + 'static,
-        F: Future<Output = ()>,
     {
         let listener = shared::create_listener_stream(&self.sock_addr, &self.server_info_file)?;
         let handler = self.svc.take().unwrap();
-        let map_svc = MapService { handler };
+
+        // Create a channel to send shutdown signal to the server to do graceful shutdown in case of non retryable errors.
+        let (internal_shutdown_tx, internal_shutdown_rx) = mpsc::channel(1);
```

Comment:
We need to store that as a variable inside the ReduceService struct for it to be used inside the reduce_fn function. Since oneshot is not cloneable, I had to opt for the mpsc method.

Reply:
please add a comment there?